### DISCUSSION TITLE
General Observations about using Moddable SDK

### DISCUSSION THREAD

**Discussion by ScreamZ** [HipsterBrown/xs-dev#184] [General]
Hi,

Here is a list of observations I caught when trying to play with xs-dev for the first time, the idea is to detect common issues, or things that could be enhanced (or to catch potential bugs).

It's given a TODO-list, so therefore we can check and pick in separated issues to improve the software !

This list will be updated continuously.

----
## Installation

- [x] Installation issue on MacOS, python related (https://github.com/HipsterBrown/xs-dev/issues/182)

## Development experience

- [x] Unable to use template (https://github.com/HipsterBrown/xs-dev/issues/183)
- [x] First draft starter kit (https://github.com/ScreamZ/xs-dev-Typescript-template)
- [x] TypeScript partially supported out-of-the-box. (https://github.com/HipsterBrown/xs-dev/issues/186)
- [ ] Sometimes xsbug do not start when running `xs-dev run`. Have to launch manually from `.local/share/moddable/build/bin/mac/release`. Maybe could be nice to add a utility command that start it?
- [ ] Once build is deployed on ESP32, Loader remains and shell too. (To be completed by @ScreamZ)
- [ ] Using Non-mac module (pins/digital) on emulator doesn't show the good error in build console, whereas using mcconfig it does.

## ESP32 Build

- [ ] Sometimes baud rate is too fast, add a paragraph in the doc to explain how to change it UPLOAD_SPEED=115200  see https://github.com/Moddable-OpenSource/moddable/issues/74

## Some typings are wrong
- [ ] Unable to import typings in a whole
- [ ] `address` is typed `addresss` in Websocket client

## Documentation improvement

- [x] Usage with typescript + ease of development with bundler. (https://github.com/HipsterBrown/xs-dev/pull/187)

**Comment by phoddie** [comment]
> Sometimes baud rate is too fast...

The optimal baud rate generally depends on the board you are using and so are defined by the host. I'm curious what "sometimes"  you are referring to here? (Always happy to have updates to improve the documentation, of course)

**Reply by ScreamZ** [reply]
Hey ! I've multiple kinds of ESP devices and I had to adjust baud rate that's true.

The thing is that if you're totally new to that, there is no indication that the failure on compilation could be solved by passing the following environment variable.

I've just suggested a paragraph in the trouble shooting section :)

**Reply by phoddie** [reply]
> The thing is that if you're totally new to that, there is no indication that the failure on compilation could be solved by passing the following environment variable.

It is a firmware flashing error, not a compilation error. There's no way to know for certain that the failure is caused by an incorrect baud rate. That makes providing a targeted error message impractical. 

> I've just suggested a paragraph in the trouble shooting section :)

That makes sense. In fact, there is a [paragraph](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/devices/esp32.md#troubleshooting-3) about it in the documentation:

> Note: Instead of modifying the make.esp32.mk file, you can make a temporary change by setting the environment variable `UPLOAD_SPEED`.

Perhaps that's not enough. What would work better for you?
